PWGHF/D2H/Tasks/taskXicToXiPiPi.cxx

```diff
@@ -432,8 +430,69 @@ struct HfTaskXicToXiPiPi {
         registry.fill(HIST("hYGenWithProngsInAcceptance"), yParticle, ptParticle);
       }
     } // gen
-  }   // process
-  PROCESS_SWITCH(HfTaskXicToXiPiPi, processMc, "Process MC", false);
+  }
+
+  void processWithDCAFitter(soa::Filtered<soa::Join<aod::HfCandXic, aod::HfSelXicToXiPiPi>> const& candidates)
+  {
+    fillHistograms(candidates);
+  }
+  PROCESS_SWITCH(HfTaskXicToXiPiPi, processWithDCAFitter, "Process Run 3", true);
+
+  void processWithKFParticle(soa::Filtered<soa::Join<aod::HfCandXic, aod::HfCandXicKF, aod::HfSelXicToXiPiPi>> const& candidates)
+  {
+    fillHistograms(candidates);
+
+    // fill KFParticle specific histograms
+    for (const auto& candidate : candidates) {
+      if (yCandRecoMax >= 0. && std::abs(candidate.y(o2::constants::physics::MassXiCPlus)) > yCandRecoMax) {
+        continue;
+      }
+      registry.fill(HIST("hChi2topoToPV"), candidate.chi2TopoXicPlusToPV(), candidate.pt());
+      registry.fill(HIST("hChi2topoXiToXicPlus"), candidate.chi2TopoXiToXicPlus(), candidate.pt());
+      registry.fill(HIST("hChi2geoXi"), candidate.kfCascadeChi2(), candidate.pt());
+      registry.fill(HIST("hChi2geoLam"), candidate.kfV0Chi2(), candidate.pt());
```

Comment:
This can be included in `fillHistograms` inside an `if constexpr (useKfParticle) {}` block, using `bool useKfParticle` as a template argument of `fillHistograms`.